### PR TITLE
66 routing system branch 4

### LINKED ISSUE #66: [IMPLEMENT] Routing System
# Overview

We need the routing system to work before we can test our main.js. It should work like it did in Lab 7. You should be able to adapt the code you wrote for Lab 7 to work in here. If you did not complete Lab 7, or cannot finish the system and get it working in a couple of hours (or have been working on it for hours and made little progress), then request help **ASAP**. 

All pages should be linked and clicking on the buttons should being up the respective pages, and those pages are already laid out (but unfinished). 

# Story points

TBD

### LINKED ISSUE #103: [DOCS] Remove Console Log Statements and Add Comments (Router.js, main.js)
Remove any print statements (console.log, etc.) from code and otherwise clean up the code with comments, from Router.js and main.js.

### PR DESCRIPTION
**Pull request by mreswick**
## Details 
- Router:
I made a large pull request with multiple things; now I'm breaking it up. This is the routing system, updated and completed. Please note that the primary changes are to main.js; Router.js is largely the same, and index.html only has a small change (with the "a" tags in navbar).

## New Changes
- previous router file updated to main's Router.js file name. Prior console.log statements here mostly removed (apart from when clicking to go to a different page, as confirmation); otherwise it's mostly the same.
- main.js updated with finished routing system. Also, comments have been added; one example is at the top of the file, which describes how to use the router (i.e., change which pages you want available to go to) so that we may add/remove pages from our website during development more easily).
- index.html updated briefly: 1. the tags under have been changed to work with SPA; a comment is given there describing this change. 2. some comments have been added to clarify the sections as pages

## Linked Issue
Include issue number here
#66

## Screenshots
Please include screenshots of any visual changes to the project (new HTML/CSS)
1. index.html: 
image
2. main.js:
Updated:
 - comments:
image
 - variable declarations/initializations at start of file
 - init(), addPageToRouterHelper(), addPagesToRouter(), bindPageToClickLink(), createArrPageLinks()
3. Router.js:
 - prior file name updated
 - old console log statements removed

## Checklist
- [X] I have requested 2+ code reviewers
- [X] I have reviewed and commented my own code.
- [ ] My code follows the style guidelines of this project.
- [ ] I have added new tests as necessary.
- [ ] All new and existing tests pass successfully.

### CODE REVIEW

**Comment by mreswick** [comment]
- extraneous files of prior pull request (#100, from branch 66-routing-system-branch_3) removed

File: source/js/Router.js [hunk 1]
```diff
@@ -5,114 +5,128 @@
...
+       *  3. Next, if statePopped is false and window.location.hash does NOT match the
+       *     hash that you just made, use history.pushState() to add the current state
+       *     and URL + hash to history
+       *  4. Finally, call the stored function for the given page
+       */
+  
+      // self-coded:
+      // 1.
+      /*
+      console.log("number of pages/properties of this: " + Object.keys(this).length);
```
Comment (RainbowPangolin) [inline source/js/Router.js:85]: Create a GitHub Issue for removing these commented out console log statements.

File: source/js/components/ExpandedRecipeCard.js [hunk 2]
```diff
@@ -1,238 +1,751 @@
...
+    // return data of current expanded recipe:
+    get data() {
+      return this.json;
     }
 
-    header p {
-      margin: 0;
-    }
+  } // end of class
+  customElements.define('recipe-expand', RecipeExpand);
```
Comment (RainbowPangolin) [inline source/js/components/ExpandedRecipeCard.js:539]: I don't think defining a custom html element is necessary.  @AntoniaY31 and @fjontran had already completed a recipe-info.html (expanded recipe page), which we can incorporate directly into index.html, and populate directly.

Comment (mreswick) [reply source/js/components/ExpandedRecipeCard.js:539]: this is already in the main branch

**Comment by mreswick** [comment]
Issue #103 added per Kaiwen's request.

File: source/js/Router.js (hunk 1 above)
Comment (mreswick) [reply source/js/Router.js:85]: sure, #103.

File: source/index.html [hunk 3]
```diff
@@ -66,6 +77,18 @@
       </div>
     </section>
 
+		<!-- Misha/Laurence added: -->
+    <section class="section--recipe-cards">
```
Comment (RainbowPangolin) [inline source/index.html:81]: What is this for? We have sections for recipe cards within each of the sections (explore, search results, etc)

File: source/styles/main.css [hunk 4]
```diff
@@ -1,105 +1,106 @@
 .hidden {
```
Comment (RainbowPangolin) [inline source/styles/main.css:1]: Just a note to any other reviewers- it looks like the only change here was auto-linting.

**Review by RainbowPangolin — COMMENTED** [review]
One more clarification question

File: source/js/components/ExpandedRecipeCard.js (hunk 2 above)
Comment (RainbowPangolin) [reply source/js/components/ExpandedRecipeCard.js:539]: Ah I see. Unfortunately, it looks like the html was already created. I'll get back to you on what we'll do with this after today's SCRUM meeting.

File: source/index.html [hunk 5]
```diff
@@ -26,32 +27,42 @@
...
-    <!-- TODO: Search bar should call Router and display search results, fetch results -->
-    <form action="./search-results.html" class="search-bar">
-      <input type="text" name="" id="" placeholder="Search recipes">
-      <button type="submit">
-        <i class="fa fa-search"></i>
-      </button>
-    </form>
-  </nav>
+	<!-- Header of website:-->
+	<section class="website-header">
```
Comment (farisashai) [inline source/index.html:42]: why are we wrapping this in a section? the `<nav>` element for the navbar is separate and outside our `<main>` tag.

File: source/index.html [hunk 6]
```diff
@@ -26,32 +27,42 @@
...
+		<nav>
+			<div class="nav-logo">
+				<img src="/source/images/pot.svg" alt="" class="nav-icon">
+				<span class="nav-title">Recipe Book</span>
+			</div>
+			<!-- TODO: Call Router -->
+      <!-- Note: the onclick="return false;" here is required for SPA/router;-->
+      <!-- -otherwise, page will reload and will get 404 error for requested link -->
+      <!-- (If don't use this, then make them buttons to avoid reload.)-->
+			<a href="" onclick="return false;" id="Home_Link">Home</a>
```
Comment (farisashai) [inline source/index.html:52]: i think it's a little cleaner to just change all of these ` <a>` tags to `<button>` elements and define event handlers in main.js for them that use the router object so we don't need to modify anything else in the navbar

File: source/index.html (hunk 3 above)
Comment (farisashai) [reply source/index.html:81]: ^ im also kinda confused. we already have a section below with class="recipe-info" for the expanded recipe where we can append the ExpandedRecipePage web component as a child when we need to load it and i dont think we should need any new sections?

File: source/js/Router.js [hunk 7]
```diff
@@ -5,114 +5,128 @@
...
-    if (typeof this[page] !== 'function') { // note to self that checking if this[page] == undefined returns true even if the function this[page] is defined
-      console.log('ERROR note to self: requested function for page: ', page, ' does not exist');
-      return;
-    }
+    navigate(page, statePopped) {
+      console.log("");
+      console.log("");
+      console.log(`navigate() function called, requested page: ${page}`);
+      /**
+       * TODO - Part 1 - Step 4
```
Comment (farisashai) [inline source/js/Router.js:66]: can we try to remove all of the unnecessary comments from the lab

File: source/js/components/ExpandedRecipeCard.js [hunk 8]
```diff
@@ -1,238 +1,751 @@
...
+      this.attachShadow({ mode: 'open' });
+  
+      // Create styles and root element
+      const styles = document.createElement('style');
+
+      // Fill in styles and root element
+      styles.innerHTML = `
+      @import url("https://fonts.googleapis.com/css2?family=Lato&display=swap");
+        
+      html,
```
Comment (farisashai) [inline source/js/components/ExpandedRecipeCard.js:16]: the styles here should match antonia and jon's design exactly from their example code in `source/demo-code/recipe-info-rough/recipe-page.css`. we should just copy all of the css from that file into the style element here.

File: source/js/components/ExpandedRecipeCard.js [hunk 9]
```diff
@@ -1,238 +1,751 @@
...
+        margin-top: .5rem;
+      }`;
+
+      // add dummy <article> wrapper element
+      // for content of custom element; leave blank:
+      let articleEl = document.createElement("article");
+      articleEl.innerHTML = ""; // empty/blank
+  
+      // Append the two <style> and <article> elements to the shadow root:
+      this.shadowRoot.append(styles);
```
Comment (farisashai) [inline source/js/components/ExpandedRecipeCard.js:235]: im not sure if it makes much of a difference but these could just go in the set data function instead? we will never be creating a recipe card without immediately populating the data

**Comment by RainbowPangolin** [comment]
Router has been implemented by @farisashai. Closing PR.

**Pull request closed without merging**